Add some validation for headers in characters

deweerdt opened the pull request:
We add strict validation for the headers that are present in the header name and the header field. We have a special case for pseudo headers, since there's a limited list of those that are allowed.

kazuho commented:
Thank you for the PR. I have a mixed impression to this. Please let me lay down my thoughts at the design level before going down to the implementation details.

i. restricting the characters used in header name

I think it is a nice thing to have. HTTP RFCs have always designated _token_ as the characters to be used in the header names, and my understanding is that people have obeyed to the suggestion.

ii. restricting the characters used in header values

I doubt if this is the right thing to do. Containing non-US-ASCII characters in header values have been traditionally permitted in HTTP and my understanding is that RFC 7230 merely obsoleted it, quote ([A. 2](https://tools.ietf.org/html/rfc7230#section-3.2.6)):
```
   Non-US-ASCII content in header fields and the reason
   phrase has been obsoleted and made opaque (the TEXT rule was
   removed).
```
Living in a Japan, I would not be surprised to see UTF-8 values stored in headers.

iii. restricting the names of psuedo-headers

I think it is already done [here](https://github.com/h2o/h2o/blob/v2.0.0/lib/http2/hpack.c#L448).

iv. how to return validation errors

The PR reuses the existing error path for reporting errors from the HPACK layer. The returned error is handled as a connection error. I think that would be fine if we are to only apply _i_. But if you are interested in restricting the header names, it _might_ be better to use a different error-reporting scheme (e.g. return 400 error) to help users identify the cause of the error, considering the fact that it is more likely to cause interoperability issues.

deweerdt commented:
> i. restricting the characters used in header name
>
> I think it is a nice thing to have. HTTP RFCs have always designated token as the characters to be used in the header names, and my understanding is that people have obeyed to the suggestion.

Ok

> ii. restricting the characters used in header values

My main motivation was to disallow `\0`, `\r` and `\n`, would only disallowing those work for you? I went the stricter route since my understanding was that we could, and being stricter tends to help security-wise. I do understand your point about interoperability.

> iii. restricting the names of psuedo-headers

Yes it is, I missed that, thanks.

> iv. how to return validation errors
> But if you are interested in restricting the header names, it might be better to use a different error-reporting scheme (e.g. return 400 error) to help users identify the cause of the error, considering the fact that it is more likely to cause interoperability issues.

Did you mean "restricting header values"? If so, I agree: using the payload in GOAWAY makes things a bit difficult to debug.

kazuho commented:
> My main motivation was to disallow `\0`, `\r` and `\n`, would only disallowing those work for you?

Yes. I wouldn't argue against prohibiting control characters (e.g. BS, DEL) as well.

> Did you mean "restricting header values"? If so, I agree: using the payload in GOAWAY makes things a bit difficult to debug.

Yes, you are right. Thank you for the correction.

kazuho commented:
Now that we have an agreement on the requirement, let's move on to how we should implement it.

To me, the optimal way to implement this is by adding a bit flags describing the type of the character to `nghttp2_huff_decode`, and use the information while decoding the encoding to collect the type of the characters used in an encoded string. For non-compressed header names and values, we could simply implement a check function (like you did).

Considering the fact that most strings (especially the long ones) would be huffman-encoded, I believe we can implement the checks practically zero-overhead if we take this path.

How does it sound?

deweerdt commented:
Hello @kazuho, it sounds good, thanks. I'll update the PR in the coming days.

kazuho commented:
FYI lib/http2/hpack_huffman_table.h is generated using https://github.com/nghttp2/nghttp2/blob/master/mkhufftbl.py. You might want to modify it (in order to mark certain characters as disallowed), and I am happy to include the modified version of the script under the `misc/` directory of H2O.

deweerdt commented:
Thanks! I had written a small C file to do that, but I also modified `mkhufftbl.py`, it helped double check everything was working as expected. I'm actually struggling a bit with sending a 400 error from `h2o_hpack_parse_headers`. At that point the stream is still in a receive state, and it sounds like neither `h2o_send_error_deferred_xxx` nor `h2o_send_error_xxx` can be called at that point.
Is the idea to set a flag in the request, and after it's all been parsed send an error? A flag that we would check in `h2o_process_requests`?

deweerdt commented:
Here's what I mean, regarding the error handling:
```
diff --git a/include/h2o.h b/include/h2o.h
index e3e9f9c..184aa59 100644
--- a/include/h2o.h
+++ b/include/h2o.h
@@ -939,6 +939,7 @@ struct st_h2o_req_t {
     h2o_ostream_t *_ostr_top;
     size_t _next_filter_index;
     h2o_timeout_entry_t _timeout_entry;
+    struct st_send_error_deferred_t *deferred_error;
     /* per-request memory pool (placed at the last since the structure is large) */
     h2o_mem_pool_t pool;
 };
diff --git a/lib/core/request.c b/lib/core/request.c
index 4f3c305..ba7aca3 100644
--- a/lib/core/request.c
+++ b/lib/core/request.c
@@ -252,6 +252,10 @@ void h2o_dispose_request(h2o_req_t *req)
 void h2o_process_request(h2o_req_t *req)
 {
     h2o_hostconf_t *hostconf = setup_before_processing(req);
+    if (req->deferred_error) {
+            req->deferred_error->_timeout.cb(&req->deferred_error->_timeout);
+            return;
+    }
     process_hosted_request(req, hostconf);
 }
 
@@ -487,12 +491,16 @@ void h2o_send_error_generic(h2o_req_t *req, int status, const char *reason, cons
         h2o_send_error_generic(args->req, args->status, args->reason, args->body, args->flags);                                    \
     }                                                                                                                              \
                                                                                                                                    \
-    void h2o_send_error_deferred_##status_(h2o_req_t *req, const char *reason, const char *body, int flags)                 \
+    void h2o_send_error_deferred_##status_(h2o_req_t *req, const char *reason, const char *body, int flags, int asap)              \
     {                                                                                                                              \
         struct st_send_error_deferred_t *args = h2o_mem_alloc_pool(&req->pool, sizeof(*args));                                     \
         *args = (struct st_send_error_deferred_t){req, status_, reason, body, flags};                                              \
-        args->_timeout.cb = send_error_deferred_cb_##status_;                                                                      \
-        h2o_timeout_link(req->conn->ctx->loop, &req->conn->ctx->zero_timeout, &args->_timeout);                                    \
+        args->_timeout.cb = send_error_deferred_cb_##status_;                                                                  \
+        if (asap){                                                                                                                 \
+            h2o_timeout_link(req->conn->ctx->loop, &req->conn->ctx->zero_timeout, &args->_timeout);                                \
+        } else {                                                                                                                   \
+            req->deferred_error = args;                                                                                            \
+        }                                                                                                                          \
     }
 
 DECL_SEND_ERROR_DEFERRED(502)
@@ -566,7 +574,7 @@ void h2o_send_redirect_internal(h2o_req_t *req, h2o_iovec_t method, const char *
     /* parse the location URL */
     if (h2o_url_parse_relative(url_str, url_len, &url) != 0) {
         /* TODO log fprintf(stderr, "[proxy] cannot handle location header: %.*s\n", (int)url_len, url); */
-        h2o_send_error_deferred_502(req, "Gateway Error", "internal error", 0);
+        h2o_send_error_deferred_502(req, "Gateway Error", "internal error", 0, 1);
         return;
     }
     /* convert the location to absolute (while creating copies of the values passed to the deferred call) */
diff --git a/lib/http2/connection.c b/lib/http2/connection.c
index 5ecbe38..5f9586d 100644
--- a/lib/http2/connection.c
+++ b/lib/http2/connection.c
@@ -317,8 +317,13 @@ static int handle_incoming_request(h2o_http2_conn_t *conn, h2o_http2_stream_t *s
 
     header_exists_map = 0;
     if ((ret = h2o_hpack_parse_headers(&stream->req, &conn->_input_header_table, src, len, &header_exists_map,
-                                       &stream->_expected_content_length, &stream->cache_digests, err_desc)) != 0)
-        return ret;
+                                       &stream->_expected_content_length, &stream->cache_digests, err_desc)) != 0) {
+        if (ret == H2O_HTTP2_ERROR_INVALID_HEADER_VALUE) {
+            h2o_send_error_deferred_400(&stream->req, "Invalid headers", "Invalid header value found", 0, 0);
+        } else {
+            return ret;
+        }
+    }
 
 #define EXPECTED_MAP                                                                                                               \
     (H2O_HPACK_PARSE_HEADERS_METHOD_EXISTS | H2O_HPACK_PARSE_HEADERS_PATH_EXISTS | H2O_HPACK_PARSE_HEADERS_SCHEME_EXISTS)
```

kazuho commented:
@deweerdt 
> I had written a small C file to do that, but I also modified `mkhufftbl.py`, it helped double check everything was working as expected.

Great!

> I'm actually struggling a bit with sending a 400 error from `h2o_hpack_parse_headers`. At that point the stream is still in a receive state, and it sounds like neither `h2o_send_error_deferred_xxx` nor `h2o_send_error_xxx` can be called at that point.

If we are only going to restrict the use of NUL,CR,LF in header values, I am perfectly fine with using GOAWAY.

As we tighten the restriction, the chance becomes higher that the rules disrupt a HTTP request; therefore we would need to use a more fine-grained error indication that describes the cause of the error. We have two options here: RST_STREAM and HTTP error response. However, since RST_STREAM cannot carry the cause of the error, HTTP error response becomes the only practical method.

If we are to use HTTP error response for the matter, I _think_ you can call `h2o_send_error_XXX` directly immediately after calling `        h2o_http2_stream_set_state(conn, stream, H2O_HTTP2_STREAM_STATE_SEND_HEADERS)`. But I might be wrong.

Please push a WIP branch so that we could understand what is exactly going on in such case, if you think we should pursue the path of using HTTP error response.

deweerdt commented:
Unfortunately, calling `h2o_http2_stream_set_state` fails with the following assert: `h2o:/include/h2o/http2_internal.h:383: void h2o_http2_stream_set_state(h2o_http2_conn_t *, h2o_http2_stream_t *, h2o_http2_stream_state_t): Assertion `stream->state == H2O_HTTP2_STREAM_STATE_REQ_PENDING' failed.`

I did push a wip version, hopefully this clarifies what i'm trying to do.

deweerdt commented:
Thinking a bit more about sending the error immediately, i don't know if that's something we can do in h2, right? For the the stream to change state, allowing us to send the response back, we need to parse the request in its entirety, don't we?

kazuho commented:
Sorry I haven't had time to reproduce the issue on my side (though I suspect first setting the state to REQ_PENDING and then to SEND_HEADERS might solve the problem).

> Thinking a bit more about sending the error immediately, i don't know if that's something we can do in h2, right? For the the stream to change state, allowing us to send the response back, we need to parse the request in its entirety, don't we?

The specification allows that. Quoting [RFC 7540 section 6.2](https://http2.github.io/http2-spec/#rfc.section.6.2): _HEADERS frames can be sent on a stream in the "idle", "reserved (local)", "open", or "half-closed (remote)" state_. An "open" stream is a stream in which the server has received the request `HEADERS` frame but not the `END_STREAM` flag.

From the implementation perspective, since HTTP/1, servers have occasionally sent error responses before receiving the entire request (in fact, 100-conitinue encourages such approach), and I _think_ it is unlilkely that doing so would cause interoperability issues in HTTP/2.

deweerdt commented:
> (though I suspect first setting the state to REQ_PENDING and then to SEND_HEADERS might solve the problem).

Thanks, that did work. I've pushed a version that sends the response directly. Let me know what you think.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -306,11 +376,23 @@ static int decode_header(h2o_mem_pool_t *pool, struct st_h2o_decode_header_resul
         const h2o_token_t *name_token;
-        if ((result->name = decode_string(pool, src, src_end)) == NULL)
+        uint8_t flags = 0;
+        if ((result->name = decode_string(pool, src, src_end, &flags)) == NULL)
```
How about  changeing arguments of `decode_string` to `h2o_mem_pool_t *pool, const uint8_t **src, const uint8_t *src_end, int is_header_name, const char **err_desc)`?

By changing as such, I believe the code would become simpler since you can do all the checks in `decode_string` function without using `H2O_HUFF_ENCODED` flag.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -306,11 +376,23 @@ static int decode_header(h2o_mem_pool_t *pool, struct st_h2o_decode_header_resul
+            if (flags & H2O_HUFF_ENCODED) {
+                if (flags & H2O_HUFF_INVALID_FOR_HEADER_NAME) {
+                    *err_desc = "found an invalid character in header name";
+                    return H2O_HTTP2_ERROR_PROTOCOL;
```
How about using HTTP response level error notification for invalid header names as well?

kazuho commented on lib/http2/hpack.c:
```diff
@@ -23,6 +23,7 @@
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <stdbool.h>
```
I would appreciate it if you could refrain from using `stdbool.h`.

We allow the header files of H2O to be included from C++ (which means that `bool` might be a C++ type), and therefore my preference is to not use `bool` in our code (but instead use `int` or `char` for the purpose) to avoid confusion.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -65,14 +66,69 @@ static h2o_iovec_t *alloc_buf(h2o_mem_pool_t *pool, size_t len)
+    };
+    int i;
+
+    if (*s == ':') {
```
Do we need to have this checked here? My understanding is that it is done in `h2o_hpack_parse_headers`.

kazuho commented:
@deweerdt Thank you for the updates!

The changes to `lib/http2/connection.c` has become so simple! I like it very much.

I've left two inline comments. Please let me know what you think (or adjust the code if they make sense).

deweerdt commented:
@kazuho thank you for your comments. I believe I addressed all of them in 74b4667

kazuho commented on lib/http2/connection.c:
```diff
@@ -317,8 +317,16 @@ static int handle_incoming_request(h2o_http2_conn_t *conn, h2o_http2_stream_t *s
     if ((ret = h2o_hpack_parse_headers(&stream->req, &conn->_input_header_table, src, len, &header_exists_map,
-                                       &stream->_expected_content_length, &stream->cache_digests, err_desc)) != 0)
+                                       &stream->_expected_content_length, &stream->cache_digests, err_desc)) != 0) {
+        if (ret == H2O_HTTP2_ERROR_INVALID_HEADER_NAME || ret == H2O_HTTP2_ERROR_INVALID_HEADER_VALUE) {
```
Do we need to distinguish between `H2O_HTTP2_ERROR_INVALID_HEADER_NAME` and `H2O_HTTP2_ERROR_INVALID_HEADER_VALUE`?

If not, let's merge them into a single error code. The detail can be found in `*err_desc` anyways.

kazuho commented on lib/http2/connection.c:
```diff
@@ -317,8 +317,16 @@ static int handle_incoming_request(h2o_http2_conn_t *conn, h2o_http2_stream_t *s
+            /* fast forward the stream's state so that we can start sending the response */
+            h2o_http2_stream_set_state(conn, stream, H2O_HTTP2_STREAM_STATE_REQ_PENDING);
+            h2o_http2_stream_set_state(conn, stream, H2O_HTTP2_STREAM_STATE_SEND_HEADERS);
+            h2o_send_error_400(&stream->req, "Invalid headers", *err_desc, 0);
```
Please uppercase `h`. Every word in the reason phrase is generally uppercased.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -154,13 +195,36 @@ static h2o_iovec_t *decode_string(h2o_mem_pool_t *pool, const uint8_t **src, con
+        if (hflags & NGHTTP2_HUFF_INVALID_FOR_HEADER_VALUE && !is_header_name) {
+            *err_desc = "found an invalid character in header value";
+            return NULL;
+        }
```
How about implementing the check using a single big `if (is_header_name) { ... } else { ... }` block? IMO doing so would improve readability as well as reduce the number of conditional branches.

Also, I think you would need to test the length of the header name before evaluating `ret->base[0]` (and return an invalid header error if the length is zero).

kazuho commented on lib/http2/hpack.c:
```diff
@@ -154,13 +195,36 @@ static h2o_iovec_t *decode_string(h2o_mem_pool_t *pool, const uint8_t **src, con
+        if (!is_header_name && contains_invalid_field_value_char((char *)*src, len)) {
+            *err_desc = "found an invalid character in header value";
+            return NULL;
+        }
```
The discussion on the huffman-encoded side applies to these blocks as well.

kazuho commented:
@deweerdt Thank you for the changes.

I've left my ideas / suggestions in the in-line comments.

Also, I think we need to record the validation errors in `st_h2o_hpack_header_table_entry_t` and return the recorded error in case the values are used (in [this block](https://github.com/h2o/h2o/blob/v2.1.0-beta1/lib/http2/hpack.c#L293-L300)).

Otherwise, repetitive use of invalid headers might pass through our validation considering the fact that we continue to receive new headers after a validation failure.

deweerdt commented:
> Also, I think we need to record the validation errors in st_h2o_hpack_header_table_entry_t and return the recorded error in case the values are used (in this block).
> 
> Otherwise, repetitive use of invalid headers might pass through our validation considering the fact that we continue to receive new headers after a validation failure.

Could you please describe how such an attach would work? If the validation fails, we don't record the values in the table, AFAIU.

deweerdt commented:
@kazuho I've done the changes you suggested, minus the change to `struct st_h2o_hpack_header_table_entry_t`, since I'm not sure how the attack would work. Thanks!

deweerdt commented:
> Could you please describe how such an attach would work? If the validation fails, we don't record the values in the table, AFAIU.

Nevermind, I think I understand now: we should actually be recording the values in the table if the encoder requested as much.

deweerdt commented:
I believe 9e0cc38 now addresses all the comments, and has the correct behavior for dynamically indexed headers.

deweerdt commented:
I believe 	bcb430b now addresses all the comments, and has the correct behavior for dynamically indexed headers.

deweerdt commented:
I believe 0c6899d now addresses all the comments, and has the correct behavior for dynamically indexed headers.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -41,6 +41,8 @@ struct st_h2o_hpack_static_table_entry_t {
 struct st_h2o_hpack_header_table_entry_t {
     h2o_iovec_t *name;
     h2o_iovec_t *value;
+    int decode_return_value; /* record the soft error for bad chars in headers */
```
I think we do not need to retain the value, since if a soft error has occurred can be found by checking the value of `err_desc`, and in such case the error code would always be `H2O_HTTP2_ERROR_INVALID_HEADER_CHAR`.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -240,7 +306,7 @@ static int decode_header(h2o_mem_pool_t *pool, struct st_h2o_decode_header_resul
 {
     int32_t index = 0;
-    int value_is_indexed = 0, do_index = 0;
+    int value_is_indexed = 0, do_index = 0, ret = 0;
```
In relation to https://github.com/h2o/h2o/pull/974/files#r71325566 I think there's no need to introduce `ret`. When reaching the bottom the function, the return code would either be `0` or `H2O_HTTP2_ERROR_INVALID_HEADER_CHAR` depending on if `*err_desc` is NULL.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -410,14 +485,18 @@ int h2o_hpack_parse_headers(h2o_req_t *req, h2o_hpack_header_table_t *header_tab
 
     while (src != src_end) {
         struct st_h2o_decode_header_result_t r;
         int ret = decode_header(&req->pool, &r, header_table, &src, src_end, err_desc);
```
I think you would need to create a local value for `err_desc` initialized to NULL, and pass pointer to the created value when calling `decode_header`, so that you could determine the cause of correctly memoize the error status in `decode_header`.

In current implementation it seems that you might mis-record errors for headers following an invalid header.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -410,14 +485,18 @@ int h2o_hpack_parse_headers(h2o_req_t *req, h2o_hpack_header_table_t *header_tab
                             const char **err_desc)
 {
     const uint8_t *src_end = src + len;
+    int parsing_ret = 0;
```
I think you could use test the value of `*err_desc` to determine the code that should be returned by this function.

deweerdt commented:
cd7d7dc has a rewritten error handling that i believe addresses your comments.

deweerdt commented:
I see, I thought this wouldn't be an issue since `*err_desc` is only set on soft errors. The suggestion you make will make the code more robust to changes in that assumption. I'll update the PR.

kazuho commented:
@deweerdt 
> I see, I thought this wouldn't be an issue since `*err_desc` is only set on soft errors.

Oh! I am so glad you pointed this out.

Actually, we _were_ setting `*err_desc` in one particular case, when an uppercase letter was found in the header name. This is specifically prohibited in RFC 7540 section 8.1.2.6, and the server _MUST_ send a PROTOCOL_ERROR when receiving such header.

So in case we find an invalid header, we need to either return a hard error in case an uppercase character existed, or return a soft error if one of other invalid character was used. In either case, `*err_desc` should be set.

As a side note, I assume the case was covered by [h2spec](https://github.com/summerwind/h2spec). Could you please run it against this PR to see if the issue is actually caught / or if any other issues exist as well?

deweerdt commented:
> Actually, we were setting *err_desc in one particular case, when an uppercase letter was found in the header name. This is specifically prohibited in RFC 7540 section 8.1.2.6, and the server MUST send a PROTOCOL_ERROR when receiving such header.
> 
> So in case we find an invalid header, we need to either return a hard error in case an uppercase character existed, or return a soft error if one of other invalid character was used. In either case, *err_desc should be set.
> 
Ah I see. That's not the case anymore, indeed.

> As a side note, I assume the case was covered by h2spec. Could you please run it against this PR to see if the issue is actually caught / or if any other issues exist as well?

I did, and you're correct, it fails on the fact that uppercase characters don't trigger a protocol error any more. I'll update the PR.

deweerdt commented:
@kazuho I've updated the PR with changes to the Huffman decoding that flag upper case characters. I've also modified the header validation to add support for the hard error case. h2spec now passes.

There are two errors reported by travis, none of which happened on my repo's build: https://travis-ci.org/deweerdt/h2o/builds/146435928

- One has to do with memcached. I suspect this might be a TOCTOU race, similar to what you described in an earlier PR.
- The other has to do with non-prioritized pushes, where we expect the non-pushed request to arrive first. I think that might also be attributed to a race that isn't related to this PR.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -109,7 +116,11 @@ static int contains_invalid_field_name_char(const char *s, size_t len)
+                *err_desc = err_found_upper_case_in_header_name;
+                return 1;
+            }
+            *err_desc = soft_err_found_invalid_char_in_header_name;
```
I think you would also need to adjust the return code as well.

Personally, I do not mind if this path was slow and/or written in two path, since it's an error path, and since it is not the _slowest_ path (I anticipate that HPACK decoder is slower). In other words, this path doesn't get frequently executed, nor is it an attractive target of a DoS attack.

deweerdt replied:
> I think you would also need to adjust the return code as well.

Is that so? In the soft error case, we want to continue parsing. Passing the right error code is handled in https://github.com/h2o/h2o/pull/974/files#diff-cdf5c2480fce6f3c41c628a00e404f32R392

kazuho replied:
Ah, you are right. Sorry for the fuss.

I think the confusion came from the fact that the name of the function (`contains_invalid_field_name_char`) pretty much asserts that it is a function that returns a boolean value (note that `contains_invaild_field_value_char` works that way). However, `contains_invalid_field_name_char` is actually returning if the validation should be stopped.

I would appreciate it if you could adjust the name (or possibly the signature of the function as well) to avoid the confusion.

Personally, I would think of changing the function to `const char *validate_header_name(const char *, size_t)` that returns `NULL` if validation succeeds, or an error string if it fails. After obtaining the error string, you could do a pointer comparison to see if it is a hard error or a soft error.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -207,24 +218,29 @@ static h2o_iovec_t *decode_string(h2o_mem_pool_t *pool, const uint8_t **src, con
             /* pseudo-headers are checked later in `decode_header` */
             if (hflags & NGHTTP2_HUFF_INVALID_FOR_HEADER_NAME && ret->base[0] != ':') {
-                *err_desc = "found an invalid character in header name";
+                if (NGHTTP2_HUFF_UPPER_CASE_CHAR) {
```
This looks like a constant if. I assume you were in hurry :-)

deweerdt replied:
Hugh :( I'll blame lack of coffee

kazuho commented:
Rerun the travis tests. They passed. However, I think there are two issues in the changes. Could you please take a look?

PS. If travis test fails, you can change commit hash of the last commit (by running `git commit --amend` without any changes added) and force push a PR. Since you are the only one updating the PR, I do not see any reason you should not do such a thing.

deweerdt commented:
> PS. If travis test fails, you can change the commit hash of the last commit (by running git commit --amend without any changes added) and force push a PR. Since you are the only one updating the PR, I do not see any reason you should not do such a thing.

Ah cool, thanks for the tip.

deweerdt commented:
ee6c5bc contains a fix for the constant condition. I'll wait for your response on the second issue before updating the PR.

kazuho commented:
@deweerdt Thank you for the changes and the comments. Aside from the discussion about `contains_invalid_field_name_char` all looks good to me.

deweerdt commented:
Thanks @kazuho, 5b65611 aims at clarifying the goal of the validation function.

kazuho commented:
Thank you very much for patiently polishing up the PR. All looks fine.
